### ISSUE TITLE
Conflicting setter definitions for property "type": com.amazonaws.services.codepipeline.model.ArtifactLocation

### ISSUE THREAD

**Issue by bschrameckSila** [aws/aws-sdk-java#1276]
Seeing this issue with both 1.11.162 and 1.11.179.

When calling a Java Lambda function from CodePipeline, and using the com.amazonaws.services.codepipeline.model.Job as the POJO for the Lambda input, the following exception occurs because the com.amazonaws.services.codepipeline.model.ArtifactLocation class has an overloaded "setType" methods (there are two "setType") methods.

    An error occurred during JSON parsing: java.lang.RuntimeException
    java.lang.RuntimeException: An error occurred during JSON parsing
    Caused by: java.io.UncheckedIOException: com.fasterxml.jackson.databind.JsonMappingException: Conflicting setter definitions for property "type": com.amazonaws.services.codepipeline.model.ArtifactLocation#setType(1 params) vs com.amazonaws.services.codepipeline.model.ArtifactLocation#setType(1 params)
    Caused by: com.fasterxml.jackson.databind.JsonMappingException: Conflicting setter definitions for property "type": com.amazonaws.services.codepipeline.model.ArtifactLocation#setType(1 params) vs com.amazonaws.services.codepipeline.model.ArtifactLocation#setType(1 params)
    at com.fasterxml.jackson.databind.deser.DeserializerCache._createAndCache2(DeserializerCache.java:269)
    at com.fasterxml.jackson.databind.deser.DeserializerCache._createAndCacheValueDeserializer(DeserializerCache.java:244)
    at com.fasterxml.jackson.databind.deser.DeserializerCache.findValueDeserializer(DeserializerCache.java:142)
    at com.fasterxml.jackson.databind.DeserializationContext.findContextualValueDeserializer(DeserializationContext.java:428)
    at com.fasterxml.jackson.databind.deser.std.StdDeserializer.findDeserializer(StdDeserializer.java:947)
    at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.resolve(BeanDeserializerBase.java:439)
    at com.fasterxml.jackson.databind.deser.DeserializerCache._createAndCache2(DeserializerCache.java:296)
    at com.fasterxml.jackson.databind.deser.DeserializerCache._createAndCacheValueDeserializer(DeserializerCache.java:244)
    at com.fasterxml.jackson.databind.deser.DeserializerCache.findValueDeserializer(DeserializerCache.java:142)
    at com.fasterxml.jackson.databind.DeserializationContext.findContextualValueDeserializer(DeserializationContext.java:428)
    at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.createContextual(CollectionDeserializer.java:164)
    at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.createContextual(CollectionDeserializer.java:25)
    at com.fasterxml.jackson.databind.DeserializationContext.handleSecondaryContextualization(DeserializationContext.java:669)
    at com.fasterxml.jackson.databind.DeserializationContext.findContextualValueDeserializer(DeserializationContext.java:430)
    at com.fasterxml.jackson.databind.deser.std.StdDeserializer.findDeserializer(StdDeserializer.java:947)
    at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.resolve(BeanDeserializerBase.java:439)
    at com.fasterxml.jackson.databind.deser.DeserializerCache._createAndCache2(DeserializerCache.java:296)
    at com.fasterxml.jackson.databind.deser.DeserializerCache._createAndCacheValueDeserializer(DeserializerCache.java:244)
    at com.fasterxml.jackson.databind.deser.DeserializerCache.findValueDeserializer(DeserializerCache.java:142)
    at com.fasterxml.jackson.databind.DeserializationContext.findContextualValueDeserializer(DeserializationContext.java:428)
    at com.fasterxml.jackson.databind.deser.std.StdDeserializer.findDeserializer(StdDeserializer.java:947)
    at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.resolve(BeanDeserializerBase.java:439)
    at com.fasterxml.jackson.databind.deser.DeserializerCache._createAndCache2(DeserializerCache.java:296)
    at com.fasterxml.jackson.databind.deser.DeserializerCache._createAndCacheValueDeserializer(DeserializerCache.java:244)
    at com.fasterxml.jackson.databind.deser.DeserializerCache.findValueDeserializer(DeserializerCache.java:142)
    at com.fasterxml.jackson.databind.DeserializationContext.findRootValueDeserializer(DeserializationContext.java:461)
    at com.fasterxml.jackson.databind.ObjectReader._findRootDeserializer(ObjectReader.java:1749)
    at com.fasterxml.jackson.databind.ObjectReader._bindAndClose(ObjectReader.java:1506)
    at com.fasterxml.jackson.databind.ObjectReader.readValue(ObjectReader.java:1102)
    Caused by: java.lang.IllegalArgumentException: Conflicting setter definitions for property "type": com.amazonaws.services.codepipeline.model.ArtifactLocation#setType(1 params) vs com.amazonaws.services.codepipeline.model.ArtifactLocation#setType(1 params)
    at com.fasterxml.jackson.databind.introspect.POJOPropertyBuilder.getSetter(POJOPropertyBuilder.java:300)
    at com.fasterxml.jackson.databind.deser.BeanDeserializerFactory.filterBeanProps(BeanDeserializerFactory.java:592)
    at com.fasterxml.jackson.databind.deser.BeanDeserializerFactory.addBeanProps(BeanDeserializerFactory.java:488)
    at com.fasterxml.jackson.databind.deser.BeanDeserializerFactory.buildBeanDeserializer(BeanDeserializerFactory.java:229)
    at com.fasterxml.jackson.databind.deser.BeanDeserializerFactory.createBeanDeserializer(BeanDeserializerFactory.java:142)
    at com.fasterxml.jackson.databind.deser.DeserializerCache._createDeserializer2(DeserializerCache.java:403)
    at com.fasterxml.jackson.databind.deser.DeserializerCache._createDeserializer(DeserializerCache.java:352)
    at com.fasterxml.jackson.databind.deser.DeserializerCache._createAndCache2(DeserializerCache.java:264)
    ... 28 more

**Comment by dagnir** [comment]
Same issue here, but for Kinesis https://github.com/aws/aws-sdk-java/issues/1226.

**Comment by shorea** [comment]
Thoughts on using @JsonIgnore annotation? Didn't help the Lambda case but could help those using their own ObjectMapper with standard configuration. Otherwise the only other option is Mixins.

**Comment by bschrameckSila** [comment]
I was actually able to work around the issue (finding a few more occurrences of this in the process) by using the `RequestStreamHandler` and our own `ObjectMapper` with mix-ins. The CodePipeline Job deserialization is actually kind of a mess, as you'll see below. I hope this helps someone, but it shouldn't be necessary to do this.

We found issues with the following classes (overloaded methods):
- `ArtifactLocation` (`setType`)
- `ActionTypeId` (`setCategory`, `setOwner`)
- `EncryptionKey` (`setType`)

We worked around the issue by creating a mix-in via an abstract class extending each problem class so that we could override the appropriate methods and add the `@JsonProperty` annotation.

For example, here is the `ArtifactLocation` fix:
```
private abstract class ArtifactLocationMixIn extends ArtifactLocation {
    @Override
    @JsonProperty("type")
    public void setType(ArtifactLocationType type) {
        super.setType(type);
    }
}
```
Then add it to the `ObjectMapper` with `.addMixIn(ArtifactLocation.class, ArtifactLocationMixIn.class)`.

Additionally, the JSON that is passed to the Lambda function is in the form:
```
{
  "CodePipeline.job": {
  ...
  }
}
```
One would expect it to just match the POJO, like so:
```
{
  "Job": {
  ...
}
```
As such, we needed to create a mix-in for the root element with the `@JsonRootName` set to "CodePipeline.job" and enable the `UNWRAP_ROOT_VALUE` feature on our `ObjectMapper`, otherwise you will get a `JsonMappingException` with a message saying that the `Root name 'CodePipeline.job' does not match expected ('Job') for type`.

**Comment by bschrameckSila** [comment]
Also, note that we did try `@JsonIgnore`, and it did work, but you will not deserialize those fields that you ignore; they will just end up as null.
```
private abstract class ArtifactLocationMixIn {
    @JsonIgnore
    abstract public void setType(String type);
}
```

I thought that maybe ignoring only the method we didn't want to use would still allow for deserialization using the other method, but that was not the case.

**Comment by shorea** [comment]
Yeah if IIRC you have to add a json ignore on the enum setter and a JsonProperty on the string setter for both serialization/deserialization to work.

**Comment by shorea** [comment]
Oh sorry I skimmed over where you mentioned Lambda. Lambda ignores any Jackson annotations so mixins looks to be the only option.

**Comment by shorea** [comment]
@bschrameckSila Can you provide a code sample?

**Comment by shorea** [comment]
@bschrameckSila let us know if you can provide the code sample. We should be able to fix this problem, just need to know how you are using the POJO to know the right fix.

**Comment by bschrameckSila** [comment]
Just use the standard Java RequestHandler with the Job object as your input and the function will start throwing errors:
```
import com.amazonaws.services.codepipeline.model.Job;
import com.amazonaws.services.lambda.runtime.Context;
import com.amazonaws.services.lambda.runtime.RequestHandler;

public class ProcessCodePipelineInvocation implements RequestHandler<Job, Void> {
    @Override
    public Void handleRequest(Job input, Context context) {
        // Do something.
        return null;
    }
}
```

**Comment by shorea** [comment]
Thanks! I'll reach out to the Lambda team to see if there is anything that can be done to fix this in their deserialization component. If not the next best workaround is to use a stream handler and use ObjectMapper directly to deserialize the POJO. If we do the stream handler approach we can add the JsonIgnore annotations to make it easier to handle without needing mixins.

**Comment by shorea** [comment]
@bschrameckSila We've added JsonIgnore annotations to the conflicting setters to avoid this problem. Lambda uses a special ObjectMapper that disables most annotations though so to take advantage of this you'll need to switch to using a Stream based handler and do the deserialization yourself.

An engineer from Lambda was kind enough to provide an example of this. Here he's using a Mixin but with the new JsonIgnore annotations added this won't be required.
https://github.com/asuresh8/Lambda-Java-Custom-Serializer/blob/master/src/main/java/com/amazonaws/lambda/runtime/events/samples/CustomHandler.java

Note that the JsonIgnore annotations will be in our next release, version 1.11.192. This might be released today, unsure if we are doing a release today.

**Comment by blaltarriba** [comment]
Hi all!

First of all, sorry to comment in a closed issue but I want to be sure I'm doing well and the issue is still there before open a new issue.

I'm using version 1.11.197 and I have the same problem with field `type`. I'm using a handler like @bschrameckSila commented above

```
import com.amazonaws.services.codepipeline.model.Job;
import com.amazonaws.services.lambda.runtime.Context;
import com.amazonaws.services.lambda.runtime.RequestHandler;

public class ProcessCodePipelineInvocation implements RequestHandler<Job, Void> {
    @Override
    public Void handleRequest(Job input, Context context) {
        // Do something.
        return null;
    }
}
```

If I understood well the last comment of @shorea with the changes made at 1.11.192 this code should work properly, right?

Thank you!

**Comment by shorea** [comment]
@blaltarriba unfortunately Lambda uses a custom ObjectMapper which does not consider the JsonIgnore annotation so to take advantage of the fix you have to use the Streaming based handler and do the deserialization yourself. An example was posted in the comment above.

**Comment by blaltarriba** [comment]
Ok, I will use stream then. Thank you @shorea

**Comment by varunnvs92** [comment]
Hi All, 
As you might know we have added jackson annotations for enum getters/setters to avoid the "Conflicting setter definitions for property" error in 1.11.167. But this has broken some customers. We decided the change was backwards incompatible and it will be reverted in the next version (1.11.208). If anyone is relying on those annotations to avoid the conflicting setter error, you will be broken by this change. There are 2 options to avoid this:
1) Upgrade to Jackson 2.7 which uses heuristic rules to resolve the proper method
2) Add MixIns for all classes (that contains enum) you want to serialize or deserialize. Then configure the object mapper with those mixins.
